src/channel_internal.c

```diff
@@ -0,0 +1,395 @@
+        if (channel_internal == NULL)
+        {
+            /*Codes_SRS_CHANNEL_INTERNAL_43_002: [ If there are any failures, channel_internal_create_and_open shall fail and return NULL. ]*/
+            LogError("Failure in THANDLE_MALLOC(CHANNEL_INTERNAL)(channel_internal_dispose)");
+        }
+        else
+        {
+            CHANNEL_INTERNAL* channel_internal_ptr = THANDLE_GET_T(CHANNEL_INTERNAL)(channel_internal);
+
+            channel_internal_ptr->lock = lock;
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_080: [ channel_create shall store given threadpool in the created CHANNEL_INTERNAL. ]*/
+            THANDLE_INITIALIZE(THREADPOOL)(&channel_internal_ptr->threadpool, threadpool);
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_084: [ channel_create shall call DList_InitializeListHead. ]*/
+            DList_InitializeListHead(&channel_internal_ptr->op_list);
+
+            /*Codes_SRS_CHANNEL_INTERNAL_43_086: [ channel_create shall succeed and return the created THANDLE(CHANNEL). ]*/
+            THANDLE_INITIALIZE_MOVE(CHANNEL_INTERNAL)(&result, &channel_internal);
+            goto all_ok;
+        }
+        srw_lock_destroy(lock);
+    }
+all_ok:
+    return result;
+}
+
+static void execute_callbacks(void* context)
+{
+    if (context == NULL)
+    {
+        LogError("Invalid arguments: void* context=%p", context);
+    }
+    else
+    {
+        CHANNEL_OP* channel_op = (CHANNEL_OP*)context;
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_145: [ execute_callbacks shall call the stored callback(s) with the result of the operation. ]*/
+        if (channel_op->pull_callback)
+        {
+            channel_op->pull_callback(channel_op->pull_context, channel_op->result, channel_op->data);
+        }
+        if (channel_op->push_callback)
+        {
+            channel_op->push_callback(channel_op->push_context, channel_op->result);
+        }
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_147: [ execute_callbacks shall perform cleanup of the operation. ]*/
+        THANDLE_ASSIGN(RC_PTR)(&channel_op->data, NULL);
+        THANDLE_ASSIGN(CHANNEL_INTERNAL)(&channel_op->channel_internal, NULL);
+
+        // copy to local reference to avoid cutting the branch you are sitting on
+        THANDLE(ASYNC_OP) temp = NULL;
+        THANDLE_INITIALIZE_MOVE(ASYNC_OP)(&temp, &channel_op->async_op);
+        THANDLE_ASSIGN(ASYNC_OP)(&temp, NULL);
+    }
+}
+
+static void cancel_op(void* context)
+{
+    CHANNEL_OP* channel_op = context;
+    CHANNEL_INTERNAL* channel_internal_ptr = THANDLE_GET_T(CHANNEL_INTERNAL)(channel_op->channel_internal);
+
+    /*Codes_SRS_CHANNEL_INTERNAL_43_134: [ cancel_op shall call srw_lock_acquire_exclusive. ]*/
+    srw_lock_acquire_exclusive(channel_internal_ptr->lock);
+    {
+        /*Codes_SRS_CHANNEL_INTERNAL_43_135: [ If the operation is in the list of pending operations: ]*/
+        if (channel_op->anchor.Flink != &channel_op->anchor && &channel_op->anchor == channel_op->anchor.Flink->Blink)
```

Comment:
>channel_op->anchor.Flink != &channel_op->anchor && &channel_op->anchor == channel_op->anchor.Flink->Blink [](http://example.com/codeflow?start=12&length=105)

why are both of these conditions necessary?

Reply:
Let the first one by C1, second one be C2.

Truth table:

C1T, C2T: Flink points to something else, and that something points back to it. Part of a non-empty list. 

C1T, C2F: Flink points to something else, that something doesn't point back to it. Cannot be part of a list.

C1F, C2TL: Flink points to itself. Not part of list.

C1F, C2F: Cannot happen.

Both conditions are needed.